django/middleware/security.py

```diff
@@ -1,8 +1,58 @@
+import base64
+import os
 import re
+from functools import partial
 
 from django.conf import settings
 from django.http import HttpResponsePermanentRedirect
+from django.middleware.constants import csp
 from django.utils.deprecation import MiddlewareMixin
+from django.utils.functional import SimpleLazyObject
+
+# TODO: Consider if we want to exclude CSP from a default set of views in DEBUG
+# mode.
+
+# TODO: Consider if we want to ship with the `@csp_exempt` decorator, to exclude
+# views from CSP headers.
+
+# TODO: Consider if we want to allow per-view configuration of CSP headers via
+# decorators, via a `@csp` decorator. This could be useful for 3rd party apps to
+# define their own CSP on their own views. There are security implications to
+# consider here, so this would need to be carefully considered.
+
+# TODO: Consider adding support for CSP hash sources, similar to nonce sources.
+
+# TODO: Consider adding a security check to warn if the CSP settings are empty
+# and no CSP headers will be sent. (docs/refs/checks.txt)
+
+
+def build_csp(config, nonce=None):
```

Comment:
investigate caching at least parts of this, like:

https://github.com/adamchainz/django-permissions-policy/blob/cde2f2a9242819882621974d13b28099d23eb6eb/src/django_permissions_policy/__init__.py#L150-L151

Python is not the fastest for string manipulation, and since the policy won't change request-to-request, it's ripe for caching.

I guess the nonce part can't be cached - that can be glued on per-request.

Reply:
With decorators now added to this PR, the method arguments can vary based on view decorator and per request, making it more challenging to cache this string. I can explore it more later once things stabilize.